hnn_core/network.py

```diff
@@ -1133,7 +1133,8 @@ def _instantiate_drives(self, tstop, n_trials=1):
                 self.external_drives[
                     drive['name']]['events'].append(event_times)
 
-    def add_tonic_bias(self, *, cell_type=None, amplitude, t0=0, tstop=None):
+    def add_tonic_bias(self, *, cell_type=None, section=None,
+                       bias_name='tonic', amplitude, t0=0, tstop=None):
```

Comment:
The `Cell.create_tonic_bias` method has 'soma' as the default section. Perhaps the default section should be 'soma' here instead of None, so that the default behavior is exposed to the user when they are interacting with the API.

Reply:
agreed!

Reply:
I did this because of test_read_configuration_json in test_io. In the json files, section is not defined so that test will fail. I thought that if section is None then it should work as before. If section = 'soma' (i.e. user knows about this new feature and sets it) then section: 'soma' will be added to cell_type_bias. In cell.py create_tonic_bias the default is section='soma'.

Reply:
what exactly is the error? It sounds like you need to modify the IO functions then to make the test pass ...

@gtdang sorry I missed this during review but the file should be called `io.py` not `hnn_io.py` as it's redundant to have "hnn" in the name and Python's modular organization means you'll say `hnn_core.io` ... consistent with what other packages do. Also the test file is then named by preprending `test_` so it should be `test_io.py` (already the case).

Reply:
It was renamed because to `hnn_io` because `io` conflicts with the python standard library `io` module. I ran into a problem because I use the python `io` for the GUI and the it would import from the hnn namespace. It doesn't need to be `hnn_io` but it cannot be `io`.

Reply:
@jasmainak  test_read_configuration_json checks whether creating a model by setting params in HNN core gives you the same result as creating a model from a json file. Section is not defined in the external bias in the json files so the test will not pass if I add section: 'soma' to the cell_type_bias dictionary.

So to make the test pass, I could choose not to add section='soma' to cell_type_bias, but then users who explicitly set section='soma' cannot look it up in their network. Which is why I thought not to add section='soma' to cell_type_bias if the user doesn't explicitly define a section (section=None), but to add it when they do.

Reply:
> It was renamed because to hnn_io because io conflicts with the python standard library io module. I ran into a problem because I use the python io for the GUI and the it would import from the hnn namespace. It doesn't need to be hnn_io but it cannot be io.

Right. But most packages in the Python ecosystem use an `io` module. That is why we have namespaces in Python (as opposed to matlab). You have the same pattern in every library, that helps the users to learn a new package more easily; and because of separate modules there are no conflicts. See:

```py
from scipy.io import savemat
from mne.io import read_raw_fif
from skimage.io import imread
```

Instead of doing:

```py
from hnn_core import io

io.read_network
```

I would recommend doing either:

```py
from hnn_core.io import read_network
```

or

```py
import hnn_core.io as hnn_io
```

Reply:
> @jasmainak test_read_configuration_json checks whether creating a model by setting params in HNN core gives you the same result as creating a model from a json file. Section is not defined in the external bias in the json files so the test will not pass if I add section: 'soma' to the cell_type_bias dictionary.

You basically have a backwards compatibility issue, right? The old json files do not have the section 'soma' defined in the cell_type_bias dictionary but the new json files created after adding bias using your function will have the section 'soma'. So in this case, you need a check when reading the file if section_name exists as a key in your cell_type_bias dictionary. If it does not, you can default it to 'soma' and emit a warning with a solution: users can either re-write the file with the correct section_type, or alternatively they can update net.biases after reading in the file. I would recommend the former since we have not settled on a method to update biases and external drives.

Reply:
> > It was renamed because to hnn_io because io conflicts with the python standard library io module. I ran into a problem because I use the python io for the GUI and the it would import from the hnn namespace. It doesn't need to be hnn_io but it cannot be io.
> 
> Right. But most packages in the Python ecosystem use an `io` module. That is why we have namespaces in Python (as opposed to matlab). You have the same pattern in every library, that helps the users to learn a new package more easily; and because of separate modules there are no conflicts. See:
> 
> ```python
> from scipy.io import savemat
> from mne.io import read_raw_fif
> from skimage.io import imread
> ```
> 
> Instead of doing:
> 
> ```python
> from hnn_core import io
> 
> io.read_network
> ```
> 
> I would recommend doing either:
> 
> ```python
> from hnn_core.io import read_network
> ```
> 
> or
> 
> ```python
> import hnn_core.io as hnn_io
> ```

You shouldn't shadow a builtin library with a top-level module. There are different methods folks use to avoid shadowing: 

1. Create a protected namespace by creating a subpackage. See [spicy.io](https://github.com/scipy/scipy/tree/main/scipy/io). Note that it's not a module. 
2. Use a different name for the module. [numpy/lib/_iotools.py](https://github.com/numpy/numpy/blob/main/numpy/lib/_iotools.py) - This is both a submodule `lib` and not naming their module `io`.

Reply:
Honestly I think the benefit of following a known pattern outweighs the potential issues due to conflict with a builtin library. Don't get too creative ...

Reply:
Btw -- what numpy does is to put the IO in the top-level of their namespace. That's why they have a private module which gives them flexibility to restructure their library.

Reply:
> Btw -- what numpy does is to put the IO in the top-level of their namespace. That's why they have a private module which gives them flexibility to restructure their library.

AND they don't use just the name`io` they use `npyio` to load the private module. In addition it's all nested within the `lib` subpackage nested within the `numpy` top-level package top-level for a protected namespace. I don't understand your point.

Reply:
My point is that the user doesn't see `_iotools` because it's a private module. They import thus:

```py
from numpy import loadtxt
```

it's in the top-level namespace. Numpy has just a few IO functions, so it makes sense to have a top-level import. MNE reads 23 different file formats, so there is a dedicated module. From the user perspective, I find it very strange to do:

```py
from hnn_core.hnn_io import read_network_configuration
```

If it's a known pattern, then one can leverage tab completion in ipython to import the module. You can decide to keep the IO as top level imports or from `hnn_core.io`. If the number of functions is limited, a top-level import may be fine and then you can call that file `_iotools.py` or something. I think my main objection in the name is putting `hnn` in it since it's redundant and hard to guess ... I can live with `_iotools.py` ...

Reply:
Yes an alternate could be naming it `_io.py` as a hidden module and then configure the `__init__.py ` to be able to load functions directly. That way users don't have to call `hnn_core.io` and call `from hnn_core import read_network_configuration`. Does that work?

Reply:
yep that works. Also, please document in `api.rst`, otherwise it's a hidden feature ...

Reply:
I'm gonna go ahead and resolve this one (see comment in conversation below)